Design.md

```diff
@@ -0,0 +1,128 @@
+
+## Job handling
+
+- Wrap the linux command in `bash`
+- Start the process using `exec.Cmd` in separate go routine
+- Utilize go channels to communicate with the go routine
+- The go routine should exit when job finishes or timeout reached or force stop requested
+
+## Library
+
+- `CreateJob(command, timeout)`
+    - Generate a unique ID
+    - Initialize a job object using the ID
+    - Save the reference to the job into memory for future access
+    - Start the job in a separate go routine
+    - Return the job object
+- `StopJob(jobID)`
+    - Fetch the job object from memory by job ID
+    - Check the current status of the job
+    - Stop the job process if it is running
+    - Send confirmation on success or, error on termination failure
+- `PurgeJob(jobID)`
+    - Fetch the job object from memory by job ID
+    - Check the current status of the job
+    - Remove the job from memory if it is stopped
+    - Return error if the job is not found or job is running
+- `GetJobStatus(jobID)`
+    - fetch the job object from memory by job ID
+    - return the details of a job along with output and error
+
+## API
+
+- `POST /create {"command": "pwd", "timeout": "1"}`
+- `GET /status/{id}`
+- `PUT /stop/{id}`
+- `PUT /purge/{id}`
+
+## CLI
+
+- `job-worker [server/client]` combined CLI for running server and client
+- `job-worker server [--port] [--cacert] [--cert] [--key]` run the API server
+- `job-worker client [create/stop/purge/status] [--address] [--cacert] [--cert] [--key]` run client to
+  communicate with the API server
+- `job-worker client create [--command] [--timeout]` request a new job
+- `job-worker client stop [--id]` stop a job
+- `job-worker client purge [--id]` remove a job from memory
+- `job-worker client status [--id]` fetch status of a job
+
+## mTLS
+
+- Use `openssl` to generate self-signed certificates
+- Use `RSA 2048` key and `sha256` digest
+- Shell script to generate certificates
+- Steps:
+    - Generate a root CA certificate (`ca.crt`, `ca.key`)
+    - Generate server certificate signed by the CA (`server.crt`, `server.key`)
+    - Generate client certificate signed by the CA (`client.crt`, `client.key`)
+
+## Authorization
+
+- Simple RBAC style authorization
+- Permissions:
+    - Read and write: Required for creating, stopping, and purging a job
+    - Read only: Required for querying status of a job
+- Two hard coded users to mock database
+    - Username `alice`: has read and write permissions
+    - Username `bob`: has read only permission
+- Pass username through client certificate using `Common Name` i.e. each user will have separate client certificate
+
+## Tests
+
+- Test the library
+- Test RBAC authorization
+- Test TLS connection
+
+## Third party libraries
+
+- [spf13/cobra](https://github.com/spf13/cobra) for CLI
+- [gorilla/mux](https://github.com/gorilla/mux) for routing
+- [mikespook/gorbac](github.com/mikespook/gorbac) for simple RBAC implementation
```

Comment:
This RBAC should be simple enough to use plain go rather than adding a dependency.

Reply:
Thank you for the suggestion. Updated the design doc.